milabench/executors.py

```diff
@@ -0,0 +1,612 @@
+        executor: Executor,
+        host: str,
+        *ssh_argv,
+        user: str = None,
+        key: str = None,
+        **kwargs,
+    ) -> None:
+        super().__init__(executor, **kwargs)
+        self.host = host
+        self.ssh_argv = [*ssh_argv]
+        self.user = user
+        self.key = key
+
+    def _find_node_config(self) -> Dict:
+        for n in self.pack.config["system"]["nodes"]:
+            if n["ip"] == self.host:
+                return n
+        return {}
+
+    def _argv(self, **kwargs) -> List:
+        del kwargs
+
+        if socket.gethostname() == self.host:
+            # No-op when executing on the main node
+            return []
+
+        node = self._find_node_config()
+        user = self.user or node.get("user", None)
+        key = self.key or node.get("key", None)
+        host = f"{user}@{self.host}" if user else self.host
+
+        argv = [
+            "ssh",
+            "-oCheckHostIP=no",
+            "-oStrictHostKeyChecking=no",
+            *self.ssh_argv,
+        ]
+        if key:
+            argv.append(f"-i{key}")
+        argv.append(host)
+        return argv
+
+
+class VoirExecutor(Executor):
+    """Execute an `Executor` through voir
+
+    Arguments:
+        executor: `Executor` to be executed
+        *voir_argv: voir command line arguments list
+        **kwargs: kwargs to be passed to the `pack.execute()`
+    """
+    def __init__(
+            self,
+            executor:Executor,
+            *voir_argv,
+            **kwargs
+    ) -> None:
+        super().__init__(
+            executor,
+            **{"setsid":True, **kwargs}
+        )
+        self.voir_argv = voir_argv
+
+    def _argv(self, **kwargs) -> List:
+        del kwargs
+        if voirconf := self.pack.config.get("voir", None):
+            hsh = md5(str(voirconf).encode("utf8"))
+            voirconf_file = (
+                self.pack.dirs.extra
+                / f"voirconf-{self.pack.tag}-{hsh.hexdigest()}.json"
+            )
+            with open(voirconf_file, "w") as f:
+                json.dump(fp=f, obj=voirconf, indent=4)
+            voir_argv = ("--config", voirconf_file)
+        else:
+            voir_argv = ()
+
+        return [
+            "voir",
+            *voir_argv,
+            *self.voir_argv,
+        ]
+
+
+class WrapperExecutor(Executor):
+    """Wrap an `Executor` with any command
+
+    Arguments:
+        executor: `Executor` to be executed
+        *wrapper_argv: command line arguments list
+        **kwargs: kwargs to be passed to the `pack.execute()`
+    """
+    def __init__(
+            self,
+            executor:Executor,
+            *wrapper_argv,
+            **kwargs
+    ) -> None:
+        super().__init__(
+            executor,
+            **kwargs
+        )
+        self.wrapper_argv = wrapper_argv
+
+    def _argv(self, **kwargs) -> List:
+        del kwargs
+        return [*self.wrapper_argv]
+
+
+# Roots
+class ListExecutor(Executor):
```

Comment:
This comment is not super important to address because it's mostly a code organization remark, but:

It doesn't really make sense to me for `ListExecutor` to have an `argv` method, and the behavior is dubious when it is the first parameter to `Executor`: for example, what does `SSHExecutor(ListExecutor(A, B))` mean? If we allow it (and we arguably shouldn't) I would argue that it should mean the same as `ListExecutor(SSHExecutor(A), SSHExecutor(B))` but that doesn't seem to be the case.

I think a better inheritance strategy would be something like this:

```
Executor (only exports the commands() method)
  |_ SingleCommandExecutor (exports argv())
  |   |_ PackExecutor (takes a Pack)
  |   |_ WrapperExecutor (takes a SingleCommandExecutor)
  |       |_ SSHExecutor
  |       |_ ...
  |_ ListExecutor (takes a list of Executor)
      |_ NJobs
      |_ ...
```

Reply:
Done